src/lib/coerce.js

```diff
@@ -408,11 +428,8 @@ exports.coerce = function(containerIn, containerOut, attributes, attribute, dflt
  * returns false if there is no user input.
  */
 exports.coerce2 = function(containerIn, containerOut, attributes, attribute, dflt) {
-    var propIn = nestedProperty(containerIn, attribute);
-    var propOut = exports.coerce(containerIn, containerOut, attributes, attribute, dflt);
-    var valIn = propIn.get();
-
-    return (valIn !== undefined && valIn !== null) ? propOut : false;
+    var out = _coerce(containerIn, containerOut, attributes, attribute, dflt);
+    return (out.src && out.inp !== undefined) ? out.val : false;
```

Comment:
Hmm, I'm not sure about this, but at this point I think the best way to handle it will be extensive jasmine testing.

I see we actually do have a test that contradicts one thing I was hoping to "fix":

https://github.com/plotly/plotly.js/blob/be708650801e24a0c9eb56bd6bdc755834cb8daa/test/jasmine/tests/lib_test.js#L781

Do we have a good reason for that behavior? It seems contrary to what we do with templates, as well as contrary to what I think the purpose of `coerce2` should be, which would be stated something like:
```
should set the default and return false if the user input is not valid
```

@archmoj do you see any issue if we make such a change? I feel like that test is locking down a bug rather than useful behavior.

(while we're at it there's a piece of copypasta in that test https://github.com/plotly/plotly.js/blob/be708650801e24a0c9eb56bd6bdc755834cb8daa/test/jasmine/tests/lib_test.js#L795-L798 - should be 2 `colOut` lines and 2 `sizeOut` lines)

So what I'd like to see is a test covering each of the valTypes used by `coerce2`, and for each one:
- valid input in the container
- valid input in the container but changing type (ie `'2'` to `2`) if such a thing exists for this valType
- invalid input in the container OR no input in the container, with:
  - no input in the template
  - invalid input in the template
  - valid input in the template
  - valid input in the template but changing type (ie `'2'` to `2`) if such a thing exists for this valType

I suspect getting this to work for *all* valTypes may not be worthwhile - especially since there are things like `colorscale` where even the default may be "fixed" by the coercion process, and `subplotid` where the default affects which input values are allowed. But we should be able to find a way to make it work for the ones we actually use for `coerce2`

Reply:
@alexcjohnson wondering what should be the result of this test?
https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/axes_test.js#L1791-L1797
Shouldn't it be 
```js
expect(yaxis.ticklen).toBe(undefined);
expect(yaxis.tickwidth).toBe(undefined);
expect(yaxis.tickcolor).toBe(undefined);
expect(yaxis.ticks).toBe('');
```
?

Reply:
@alexcjohnson this PR is ready for another review.
Please see https://github.com/plotly/plotly.js/pull/4904/commits/adc36cd1b04bd097f5de7ae9a694df8712e0365c.